rfd/0105-headless-authentication.md

```diff
@@ -0,0 +1,406 @@
+
+This solves two problems: 1) multiple users on the same remote machine will not have overlapping credentials on disk in `~/.tsh` and 2) attackers will not be able to steal a user's credentials from disk.
+
+We will also lock the process memory using [`unix.Mlockall`](https://pkg.go.dev/golang.org/x/sys/unix#Mlockall) to ensure that an attacker can not swap out process memory to disk, in order to extract the client's certificates.
+
+2. **Issued certificates through headless authentication are short lived (1 minute TTL)**
+
+Since the user certificates are only meant to last for a single client lifetime, the Auth server should issue the certificates with a 1 minute TTL. This limits the damage potential if an attacker manages to compromise the user's certificate, whether through a faulty client or a clever attack.
+
+Note: SSH sessions started with these certs will also be subject to the `max_session_ttl` role option.
+
+3. **User must complete WebAuthn authentication for each individual headless `tsh` request (`tsh ls`, `tsh ssh`, etc.)**
+
+Headless authentication, like any login mechanism, can be started by any unauthenticated user. To prevent phishing attacks, we must prompt the user to acknowledge and approve each headless authentication request with WebAuthn, since WebAuthn provides strong protection against phishing attacks. Legacy OTP MFA methods will not be supported.
+
+4. **The Server must verify that the client requesting certificates is the only client that can receive/use the certificates**
+
+To ensure the requesting client is the only client that can use resulting headless certificates, we will use the PKCE-like flow common to all Teleport login flows. The client provides their public key in the login request, and the server issues certificates for the client's public key. If the client does not have access to the corresponding client private key, they cannot use the certificates.
+
+Note: The server can also encrypt the certificates with the client's public key. Since we don't currently do this for other login endpoints, we will omit this from the initial design.
+
+Additionally, headless login will be handled by the Teleport Proxy endpoint - `/webapi/login/headless`. In this endpoint, the Proxy will make multiple Auth server API requests to initiate headless authentication and receive certificates. To prevent attackers from hijacking a headless auth request directly from the Auth server, The Proxy will generate a token for the auth request and pass the token in the initial headless auth request creation. The Auth server will keep this token a secret and authorize the latter API requests using this token.
```

Comment:
Since the proxy already has the public key from the client do we need an additional token? Cannot we ask for a signature or something to assert that the client is trusted?

Reply:
I added the token to somewhat mirror the normal SSO login flow where:
 - the client does SSO
 - the SSO provider generates an SSO auth request with a state token 
 - the SSO provider returns the state token to the client
 - the client receives certificates by presenting the state token

In Headless Authentication, the client is the Proxy Server (from `/webapi/login/headless`) and the "SSO" provider is the Auth server. The Proxy creates the HeadlessAuthentication and later provides the token to receive certificates. (I suppose it would be more accurate if `CreateHeadlessAuthentication` returned a state token rather than accepting one). This ensures that a third party (with permission to `GenerateUserCerts` for the user) can't take the headless authentication request ID and generate certs for it. 

Since the certs still require the client's private key with the PKCE-like flow, this isn't strictly necessary, so we could remove this.

> Cannot we ask for a signature or something to assert that the client is trusted?

This isn't something we currently do for other login endpoints, nor do we need to to maintain a PKCE-like flow. It also would not be trivial to introduce here. The endpoint transport is unauthenticated and does not have a way to challenge the client for their private key. Additionally, the Proxy service is what calls `GenerateUserCerts`, so for the Auth server to challenge the user's private key, it would need to be forwarded through the Proxy.

Reply:
> Since the certs still require the client's private key with the PKCE-like flow, this isn't strictly necessary, so we could remove this.

After further consideration I agree with you, I am not sure if this token is really buying us much.  @Joerger let me know if I am missing anything, but I believe the token was introduced to try and enforce we are talking to the same client consistently (as discussed using an alternate strategy here: https://github.com/gravitational/teleport/pull/21005#discussion_r1097916751).

This token does provide an additional secret that would need to be compromised.  However it's hard to imagine any scenario where this token would remain secure, but the clients in-memory private key could be exposed.

I think it would be better to rely on the certificate security, specifically such that:
1. This control becomes the primary enforcement:

> The client provides their public key in the login request, and the server issues certificates for the client's public key. If the client does not have access to the corresponding client private key, they cannot use the certificates.

It sounds like we have an existing pattern we can use here?  I was suggesting earlier we encrypt the responding certificate using the public key to enforce private key ownership.  However it sounds like we are using a different mechanism?  Are we chaining the certificates or can you help fill me in the details for how this is enforced?

2. `GenerateUserCerts` currently requires the token for authentication, but also the previous state to be established.  I believe the token auth can be removed as long as we are careful about making sure that established state is exactly as expected (ie setup and not already used).

TL:DR; If we can outline more about how the certificate generation is done, I feel comfortable removing this token from the spec.

Reply:
@mbaraniak-doyensec Any comments?

Reply:
I'm going to update the Design to have a single Proxy -> Auth call, and have all the authentication logic on the Auth server. This removes any privileges in the headless flow from the Proxy server and into the hands of the Auth server (as they should be). This in turn removes the need for the request token.

> It sounds like we have an existing pattern we can use here? I was suggesting earlier we encrypt the responding certificate using the public key to enforce private key ownership. However it sounds like we are using a different mechanism? Are we chaining the certificates or can you help fill me in the details for how this is enforced?

Yes, the certificates we provide to the logged in user chains their private key to the Auth Server CA, and we use that to authenticate/authorize the user in future calls. This should be enough to secure the authentication endpoint, anything more (encrypting the certs, etc.) is not necessary for security from a cryptographic perspective.

Reply:
That is true as long as the chain is correctly verified (Relevant CWE https://cwe.mitre.org/data/definitions/296.html).  But I still feel perfectly comfortable with this change (just highlighting the already existing need to validate the chain correctly).

I will try to monitor for your update, thank you!

Reply:
I have updated the RFD as mentioned.